### DISCUSSION TITLE
what´s the correct steps to apply a custom bbapend to linux-tegra

### DISCUSSION THREAD

**Discussion by danny1209** [OE4T/tegra-demo-distro#246] [General]
Hi, 

I made some changes to the files in workspace/sources/linux-tegra, everything works fine. 
Then the changes are saved as patches by using `devtool finish ... `in my own layer.
![image](https://user-images.githubusercontent.com/46848895/216622997-c482acba-a570-4751-8db4-256537693c90.png)

Now I want to rebuild all the images with the patches in own layer from very beginning, I thought the patches will be applied automaticlly during the building, but they can not be applied:
![image](https://user-images.githubusercontent.com/46848895/216624118-a4787548-df4c-4b23-ba0f-2be5e7c8a922.png)
![image](https://user-images.githubusercontent.com/46848895/216624232-71ec3605-68e4-4a00-b40e-b6fd76a90c00.png)

Any help is appreciated.

**Comment by ichergui** [comment]
Hey @danny1209 

You need to make sure that you added the patch/patches in you `*.bbappend` recipe. 

Something like the following: 

```
SRC_URI:append = " \
    file://patch1.patch \
    file://patch2.patch \
"
```

**Reply by madisongh** [reply]
@ichergui I think `devtool` handles that bit for you.

@danny1209 Could you attach the actual log file, or extract the text of those errors, rather than posting screen captures?  Information is getting cut off in the pictures that would be useful to see.

That said, I don't often use `devtool` for kernel recipes.  In theory it should work, but I'd like to see exactly why the patches aren't getting applied here.

**Reply by danny1209** [reply]
> @ichergui I think `devtool` handles that bit for you.
> 
> @danny1209 Could you attach the actual log file, or extract the text of those errors, rather than posting screen captures? Information is getting cut off in the pictures that would be useful to see.
> 
> That said, I don't often use `devtool` for kernel recipes. In theory it should work, but I'd like to see exactly why the patches aren't getting applied here.

Log data follows:
| DEBUG: Executing python function extend_recipe_sysroot
| NOTE: Direct dependencies are ['/home/user/tegra-demo-distro/layers/meta/recipes-kernel/kern-tools/kern-tools-native_git.bb:do_populate_sysroot', 'virtual:native:/home/user/tegra-demo-distro/layers/meta/recipes-devtools/patch/patch_2.7.6.bb:do_populate_sysroot', '/home/user/tegra-demo-distro/layers/meta/recipes-devtools/quilt/quilt-native_0.66.bb:do_populate_sysroot']
| NOTE: Installed into sysroot: ['patch-native', 'gnu-config-native', 'libtool-native', 'attr-native', 'autoconf-native', 'automake-native', 'texinfo-dummy-native', 'gettext-minimal-native', 'm4-native']
| NOTE: Skipping as already exists in sysroot: ['kern-tools-native', 'quilt-native']
| DEBUG: sed -e 's:^[^/]*/:/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/recipe-sysroot-native/:g' /home/user/tegra-demo-distro/build/tmp/sysroots-components/x86_64/gnu-config-native/fixmepath /home/user/tegra-demo-distro/build/tmp/sysroots-components/x86_64/libtool-native/fixmepath /home/user/tegra-demo-distro/build/tmp/sysroots-components/x86_64/autoconf-native/fixmepath /home/user/tegra-demo-distro/build/tmp/sysroots-components/x86_64/automake-native/fixmepath | xargs sed -i -e 's:FIXMESTAGINGDIRTARGET:/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/recipe-sysroot:g; s:FIXMESTAGINGDIRHOST:/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/recipe-sysroot-native:g' -e 's:FIXME_PSEUDO_SYSROOT:/home/user/tegra-demo-distro/build/tmp/sysroots-components/x86_64/pseudo-native:g' -e 's:FIXME_HOSTTOOLS_DIR:/home/user/tegra-demo-distro/build/tmp/hosttools:g' -e 's:FIXME_PKGDATA_DIR:/home/user/tegra-demo-distro/build/tmp/pkgdata/jetson-nano-devkit-emmc:g' -e 's:FIXME_PSEUDO_LOCALSTATEDIR:/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/pseudo/:g' -e 's:FIXME_LOGFIFO:/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/temp/fifo.130209:g'
| DEBUG: Python function extend_recipe_sysroot finished
| DEBUG: Executing shell function do_patch
| (1/1) 0001-linux-tegra-patches.patch
| [INFO]: check of .kernel-meta//patches//./0001-linux-tegra-patches.patch with "git am" did not pass, trying reduced context.
| [INFO]: Context reduced git-am of .kernel-meta//patches//./0001-linux-tegra-patches.patch with "git am" did not work, trying "apply".
| /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source/.git/rebase-apply/patch:1766: trailing whitespace.
| 
| /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source/.git/rebase-apply/patch:1932: trailing whitespace.
| 
| /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source/.git/rebase-apply/patch:1938: trailing whitespace.
| 
| /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source/.git/rebase-apply/patch:1951: trailing whitespace.
| 
| /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source/.git/rebase-apply/patch:1967: trailing whitespace.
| 
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0000-a01.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0000-a01.dtsi: patch does not apply
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0002-a02.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0002-a02.dtsi: patch does not apply
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0002-b00.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-gpio-p3448-0002-b00.dtsi: patch does not apply
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0000-a01.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0000-a01.dtsi: patch does not apply
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0002-a02.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0002-a02.dtsi: patch does not apply
| error: patch failed: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0002-b00.dtsi:1
| error: nvidia/platform/t210/porg/kernel-dts/porg-platforms/tegra210-porg-pinmux-p3448-0002-b00.dtsi: patch does not apply
| [ERROR]: Application of .kernel-meta//patches//./0001-linux-tegra-patches.patch failed.
|          Patch needs to be refreshed. Sample resolution script:
|              .git/rebase-apply/resolve_rejects
| ERROR: Could not apply patches for jetson-nano-devkit-emmc.
| ERROR: Patch failures can be resolved in the linux source directory /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source)
| WARNING: exit code 1 from a shell command.
| ERROR: Execution of '/home/user/tegra-demo-distro/build/tmp/work/jetson_nano_devkit_emmc-oe4t-linux/linux-tegra/4.9.299+gitAUTOINC+ff65a92ec5-r0/temp/run.do_patch.130209' failed with exit code 1
ERROR: Task (/home/user/tegra-demo-distro/layers/meta-tegra/recipes-kernel/linux/linux-tegra_4.9.bb:do_patch) failed with exit code '1'
NOTE: Tasks Summary: Attempted 1427 tasks of which 0 didn't need to be rerun and 1 failed.
Summary: 1 task failed:
  /home/user/tegra-demo-distro/layers/meta-tegra/recipes-kernel/linux/linux-tegra_4.9.bb:do_patch
Summary: There was 1 WARNING message shown.
Summary: There were 3 ERROR messages shown, returning a non-zero exit code.

**Reply by danny1209** [reply]
> Hey @danny1209
> 
> You need to make sure that you added the patch/patches in you `*.bbappend` recipe.
> 
> Something like the following:
> 
> ```
> SRC_URI:append = " \
>     file://patch1.patch \
>     file://patch2.patch \
> "
> ```
![image](https://user-images.githubusercontent.com/46848895/216633479-b6a5c69d-7fbc-4fee-aa51-4b4e0a61819b.png)

**Reply by dwalkes** [reply]
I'm guessing somehow your patches are built on a different srcrev than the one you are trying to apply to.  I'd browse into the path mentioned in the error message ```| ERROR: Patch failures can be resolved in the linux source directory /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source)```  - take a look at the source there, try hand applying the patches and see if you can figure out what's different.  Make sure your patches are built on top of the SRCREV corresponding to the kernel image you are using.  So for instance, if you are using the dunfell branch make sure you are on [this branch and SRCREV](https://github.com/OE4T/meta-tegra/blob/c6f1f512c5e21bd6290bc5f0aa4314b1be36c3a6/recipes-kernel/linux/linux-tegra_4.9.bb#L22-L23) when generating your patches.

**Reply by madisongh** [reply]
I think the problem in this case stems from the `trailing whitespace` warnings.  For most recipes, such warnings aren't fatal, but the Yocto kernel build tools use `git rebase` to apply patches, and it's apparently more sensitive to whitespace-related conflicts.  If you can update your patches to apply cleanly, without those changes, it should work better.  I just ran a couple of quick tests myself, and my patches were applied as expected.

**Reply by danny1209** [reply]
> I'm guessing somehow your patches are built on a different srcrev than the one you are trying to apply to. I'd browse into the path mentioned in the error message `| ERROR: Patch failures can be resolved in the linux source directory /home/user/tegra-demo-distro/build/tmp/work-shared/jetson-nano-devkit-emmc/kernel-source)` - take a look at the source there, try hand applying the patches and see if you can figure out what's different. Make sure your patches are built on top of the SRCREV corresponding to the kernel image you are using. So for instance, if you are using the dunfell branch make sure you are on [this branch and SRCREV](https://github.com/OE4T/meta-tegra/blob/c6f1f512c5e21bd6290bc5f0aa4314b1be36c3a6/recipes-kernel/linux/linux-tegra_4.9.bb#L22-L23) when generating your patches.

Hi @dwalkes ,
the patches are generatd and used on the same branch 
![image](https://user-images.githubusercontent.com/46848895/216644606-ee9829ab-2084-4d0d-ad21-0a4365938867.png)

**Reply by danny1209** [reply]
> I think the problem in this case stems from the `trailing whitespace` warnings. For most recipes, such warnings aren't fatal, but the Yocto kernel build tools use `git rebase` to apply patches, and it's apparently more sensitive to whitespace-related conflicts. If you can update your patches to apply cleanly, without those changes, it should work better. I just ran a couple of quick tests myself, and my patches were applied as expected.

I manually resolved the warnings about `trailing whitespace`. But the rest are still there:
![image](https://user-images.githubusercontent.com/46848895/217003062-73b25432-4dee-4a50-81a0-c8525b11395a.png)

**Reply by madisongh** [reply]
Well, it's obviously still not happy about whatever changes you made to those `dtsi` files.  You'll need to do some digging down in the kernel build workspace to see why it's failing.

**Reply by danny1209** [reply]
> Well, it's obviously still not happy about whatever changes you made to those `dtsi` files. You'll need to do some digging down in the kernel build workspace to see why it's failing.

I created the patches by these steps:
![image](https://user-images.githubusercontent.com/46848895/217207332-6ba48c18-d186-4ba7-8182-60e37cdd2d6a.png)

could you share me how you created your patch files? maybe I did something wrong.

**Reply by madisongh** [reply]
I used my usual git workflow of `git add -u` (or `git add` with specific path when adding a new file), reviewing the changes with `git diff --cached`, then separately using `git commit` to commit.  Then the usual `devtool finish linux-tegra meta-tegrademo`.

**Reply by danny1209** [reply]
Hi @madisongh 
I think I find something special.
![image](https://user-images.githubusercontent.com/46848895/217639255-01e77d2a-d3f8-4a03-939b-af1911d29b7f.png)
All these 8 files have a different format with **`CRLF line ending`**, they are different from the others. I think this is the reason why the patch can not be applied in these dts files. 
could you please confirm it? Thanks in advanced!

**Reply by madisongh** [reply]
Ah, yes, I see. It looks like `devtool`'s patch generator is messing with the CRLF line terminators.  A workaround that worked for me was to manually do `git format-patch HEAD~ -o /path/to/my/layer/recipes-kernel/linux/linux-tegra/` to have git regenerate the patch after doing the `devtool finish`.  `git format-patch` preserved the line terminators (which makes the patch look strange, but works).

`devtool` also uses `git format-patch` under the hood, but it looks like it does some additional processing on the generated patches that must not be preserving the CR characters.

**Reply by danny1209** [reply]
After doing `**devtool finish...**`, I get my regular patch file.
Then **`git format-patch HEAD~ -o /path/to/my/layer/recipes-kernel/linux/linux-tegra/`**
I get another patch file: 
![image](https://user-images.githubusercontent.com/46848895/217761516-b9f2cefc-7ba0-4517-8f30-77a3f84d414f.png)
It seems has nothing to do with my regular patch.

After that I run `**bitbake demo-image-full**`, and I get the same errors as before
![image](https://user-images.githubusercontent.com/46848895/217761911-877af262-eb69-4b05-8d09-8b42d4612741.png)
It´s not working for me. Is it possible simply update the sourcecode of these 8 files?

**Reply by danny1209** [reply]
no, this time only 6 files, seems 2 files less than before.

**Reply by madisongh** [reply]
From the name of that patch, it sounds like you were not in the `workspace/sources/linux-tegra` directory when you ran `git format-patch`.  The goal with that command is to recreate the patch that `devtool` created, so you have to be in the directory where `devtool` put its git repository for the source code, in order for it to work.  The patch should get generated with the same name as the one that got created by `devtool`.

**Reply by danny1209** [reply]
Yes, it seems working. CRLF will be replaced by LF.
but why I cant upload the updated patch file in remote repo. it shows me `The file will have its original line endings in your working directory.`

**Reply by madisongh** [reply]
Does it fail, or just give you a warning about it?  There are some `git` settings that control how it handles the mix of Unix and DOS line terminators.  If it's failing to let you submit the change, try `git config --global core.safecrlf false` to disable that, or use `warn` instead of `false` to keep the warning message but still the the change through.

**Reply by danny1209** [reply]
![image](https://user-images.githubusercontent.com/46848895/217848746-02207f22-92c9-4d92-b145-b47cee51e732.png)
this is what happens

**Reply by danny1209** [reply]
Let me make it clearly:
after `git format-patch`, I do get a new patch, which is working properly.
And I want to push this new patch in my remote repo and reuse it later. 
Then I realise the patch I pulled from repo is different from my original patch. 
Also get the same errors if I build the code with the pulled patch from remote

**Reply by madisongh** [reply]
It sounds like some kind of configuration, either locally, on the remote, or in a `.gitattributes` file in your repo, might be forcing the line ending conversions on you.  Searching around for this topic will turn up lots of information about this kind of problem with git.  I found [this article](https://www.aleksandrhovhannisyan.com/blog/crlf-vs-lf-normalizing-line-endings-in-git/), which looks pretty comprehensive.  What you want is to make sure everything is configured to ensure that existing line endings are preserved in your working tree, rather than normalizing all files to just LF termination.

**Reply by danny1209** [reply]
![image](https://user-images.githubusercontent.com/46848895/217857283-0b021532-8d53-446f-9e47-e64ba029f85a.png)
Ok, resolved by this.
==========================================
Thank you guys for all your help!!!

**Reply by ichergui** [reply]
Happy that you resolve it finally
